src/python/pants/backend/go/goals/test.py

```diff
@@ -121,8 +139,36 @@ def transform_test_args(args: Sequence[str]) -> tuple[str, ...]:
 async def run_go_tests(
     field_set: GoTestFieldSet, test_subsystem: TestSubsystem, go_test_subsystem: GoTestSubsystem
 ) -> TestResult:
-    maybe_pkg_info = await Get(
-        FallibleFirstPartyPkgInfo, FirstPartyPkgInfoRequest(field_set.address)
+    maybe_pkg_info, wrapped_target, test_sources = await MultiGet(
+        Get(FallibleFirstPartyPkgInfo, FirstPartyPkgInfoRequest(field_set.address)),
+        Get(WrappedTarget, Address, field_set.address),
+        Get(SourceFiles, SourceFilesRequest([field_set.sources])),
+    )
+    target = wrapped_target.target
+    test_sources_dirpath = field_set.address.spec_path
+
+    explicit_dependencies = await Get(
+        ExplicitlyProvidedDependencies, DependenciesRequest(target[Dependencies])
+    )
+
+    # TODO: Consider ignores as well.
+    included_targets = await Get(Targets, Addresses(explicit_dependencies.includes))
+
+    dependency_source_files = await Get(
+        SourceFiles,
+        SourceFilesRequest(
+            [tgt[SourcesField] for tgt in included_targets if tgt.has_field(SourcesField)]
+        ),
+    )
+
+    subsetted_dependency_source_files_digest = await Get(
+        Digest,
+        DigestSubset(
+            dependency_source_files.snapshot.digest, PathGlobs([f"{test_sources_dirpath}/**"])
+        ),
+    )
+    stripped_dependency_source_files_digest = await Get(
+        Digest, RemovePrefix(subsetted_dependency_source_files_digest, test_sources_dirpath)
     )
```

Comment:
The intent here is to flatten the sandbox so that the test binary and everything are at the root of the sandbox. This could also be done by moving the test binary into the actual source subdirectory and setting working_directory appropriately. Thoughts?

Reply:
I think we need to support `files` generically, meaning that you can load files loaded in an ancestor directory. This [blog](https://dave.cheney.net/2016/05/10/test-fixtures-in-go) explains:

> First, when you run go test, for each package in scope, the test binary will be executed with its working directory set to the source directory of the package under test.

So sounds like yeah, that's a good change to make. I'll do it in a precursor.

Reply:
Good find. Yeah so let's avoid flattening then.